_posts/2019-07-03-get-to-most-out-of-the-linker-map-file.md

```diff
@@ -0,0 +1,234 @@
+                    0x0000000000001240                _start
+     .text          0x00000000000012b4       0x3c _build/nrf52840_xxaa/gcc_startup_nrf52840.S.o
+                    0x00000000000012b4                Reset_Handler
+                    0x00000000000012dc                NMI_Handler
+    
+    [...]
+    
+     .text.**bsp_board_led_invert**
+                    0x00000000000012f0       **0x34** _build/nrf52840_xxaa/**boards.c.o**
+                    0x00000000000012f0                bsp_board_led_invert
+```
+
+Scrolling those lines give the address of each function and its size. Above, you can read the address of `bsp_board_led_invert`, coming from `boards.c.o` (compilation unit of `board.c` as you guessed) which as a size of 0x34 bytes in the `text` area. That way, we are able to locate each function used in the program.
+
+My constant string `_delay_ms_str` is obviously included in the program as it is initialized. Read-only data are saved as `rodata` and kept in the `FLASH` region as per my linker script (stored in Flash and not copied in RAM as it is constant). I can find it under that line:
+
+```
+    .rodata.main.str1.4
+    	              0x00000000000017b8        0x4 _build/nrf52840_xxaa/main.c.o
+```
+
+I also noticed that the inclusion of `_ctype_` added 0x101 bytes of read-only data in the `text` area 😯.
+
+```
+    .rodata._ctype_
+                    0x00000000000017c0      **0x101** /usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc_nano.a(lib_a-ctype_.o)
+                    0x00000000000017c0                **_ctype_**
+```
+
+Including a call to `atoi` made the `text` area to be increased but also the `data` area. I used a diff tool to make things simpler and discovered data that was before discarded by the linker:
+
+```
+    .data._impure_ptr
+    								0x0000000000000000        0x4 /usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc_nano.a(lib_a-impure.o)
+```
+
+One last piece of information to keep in mind: initialized variables have to be kept in Flash but they appear in `RAM` in the map file because they are copied into RAM before entering the `main` function. It means that you better have to initialize variables to `0` to put them in the `BSS` section in order to free up some Flash. Here, symbols `__data_start__` and `__data_end__` keep track of the area used in RAM to keep initialized variables. Those values are stored starting at `0x00000000000018d0`:
+
+```
+    .data           0x0000000020000008       0x70 load address 0x00000000000018d0
+                    0x0000000020000008                __data_start__ = .
+    [...]
+    								0x0000000020000078                __data_end__ = .
+```
+
+Note that symbols that are compiled but not needed are also listed in the map file in the `Discarded input sections` part.
```

Comment:
Would be good to reproduce the "discarded input sections" part of the map file here, and discuss *why* it is that those sections would get discarded.